compilers/concrete-compiler/compiler/lib/Dialect/FHE/Analysis/MANP.cpp

```diff
@@ -915,6 +950,117 @@ class MANPAnalysis
     }
   }
 
+  // Provide a mechanism to have mlir::Value as a key for a map structure
+  struct ValueHasher {
+    std::size_t operator()(const mlir::Value &v) const {
+      return mlir::hash_value(v);
+    }
+  };
+
+  // Compute the MANP value of a linalg.generic operation by emulating its
+  // execution
+  std::optional<llvm::APInt>
+  emulateLinalgGenric(mlir::linalg::GenericOp genericOp,
+                      llvm::ArrayRef<const MANPLattice *> operandMANPs) {
+    assert(genericOp.getOutputs().size() == 1 &&
+           "MANP doesn't support linalg.genric with more than one output");
+
+    // Constant output maps require accumulating the MANP values from one
+    // iteration to the other
+    auto isOutputMapConst =
+        genericOp.getIndexingMapsArray().back().isConstant();
+
+    // We want to use a different mechanism to store MANP values than the
+    // Analysis. We don't want to write anything to the lattice values
+    // controlled by the analysis, but we will use them to read values that we
+    // don't yet have in the emulation
+    std::unordered_map<mlir::Value, MANPLatticeValue, ValueHasher> valueToManp;
+    auto fetchOrFallbackToAnalysis =
+        [&](mlir::Value value) -> MANPLatticeValue * {
+      auto elem = valueToManp.find(value);
+      if (elem != valueToManp.end()) {
+        return &elem->second;
+      }
+      auto manp = getLatticeElement(value)->getValue();
+      valueToManp[value] = manp;
+      return &valueToManp[value];
+    };
+    auto loopRange =
+        mlir::concretelang::fhe::utils::getLinalgGenericLoopRange(genericOp);
+    auto iterCount = std::accumulate(loopRange.begin(), loopRange.end(), 1,
+                                     std::multiplies<int64_t>());
+    llvm::SmallVector<int64_t> strides;
+    auto stride = iterCount;
+    for (size_t i = 0; i < loopRange.size(); i++) {
+      stride /= loopRange[i];
+      strides.push_back(stride);
+    }
+
+    // init block arguments' MANP: map block arguments with op operands
+    for (auto arg : genericOp.getBlock()->getArguments()) {
+      auto argIndex = arg.getArgNumber();
+      auto operandRange = genericOp.getInputs();
+      if (argIndex >= operandRange.size()) {
+        argIndex -= operandRange.size();
+        operandRange = genericOp.getOutputs();
+      }
+      valueToManp[arg] = getLatticeElement(operandRange[argIndex])->getValue();
+    }
+
+    // indices at a specific iteration
+    llvm::SmallVector<int64_t> indices(loopRange.size(), 0);
+    std::optional<llvm::APInt> result = {};
+    for (auto i = 0; i < iterCount; i++) {
```

Comment:
I'm wondering if there is a faster way to compute this. Instead of walking the body `iterCount` times, wouldn't it be sufficient to just calculate the impact of the operations of the body on the produced result and then apply this to the MANP values from the block arguments? For parallel iterators, this should be very straightforward, as you can calculate the final MANP just from the inputs and the operations in the body. For reduction dimensions this should only be slightly more complex, as you would need to take into account the number of times the reduction is performed.

Reply:
The MANP analysis as is (with the initialization of block argument) is able to compute the MANP value for parallel loops. I tried to come up with something that don't have to iterate, but considering the fact that linalg.generic don't know the formulas for computing the MANP of the different possible operations, knowing the reduction dimension doesn't help much, as we don't know their effect when ran multiple times

Reply:
Anyway we need to have the actual plaintext values for each iterations to be able to compute an accurate manp, even with parallel iterators. I'm not sure what are you saying by "calculate the impact"?